File: Front/TS/src/domain/features/CompleteBetHandler.ts
import { IBetRepository } from "../bets/IBetRepository";

export class CompleteBetHandler {
    Handle(request: ICompleteBetRequest): Promise<void> {
        if(this.HasNeedProof(request.isSuccess, request.proof)) {
            return Promise.resolve();
        }
        this.betRepository.CompleteAsync(request.betId, request.isSuccess, request.proof)
        this.outputPort.Success({
            BetId: request.betId,
            IsSuccess: request.isSuccess
        });
        return Promise.resolve()
    }
    
    private HasNeedProof(isSuccess: boolean, proof: string | undefined) {
        if(isSuccess && !proof) {
            this.outputPort.ProofIsRequired();
            return true;
        }
        return false; 
    }
    constructor(private readonly betRepository: IBetRepository, private readonly outputPort: ICompleteBetOutputPort) {}
}

export interface ICompleteBetRequest { 
    betId: string; 
    isSuccess: boolean; 
    proof?: string; 
}

export interface ICompleteBetOutputPort {
    ProofIsRequired(): void;
    Success(response: CompleteResponse): void;
}

export interface CompleteResponse {
    BetId: string;
    IsSuccess: boolean;
}

File: Front/.Net Maui/BetFriends.Domain/Features/CompleteBet/CompleteBetHandler.cs
using BetFriends.Domain.Bets;
using MediatR;

namespace BetFriends.Domain.Features.CompleteBet;

public sealed class CompleteBetHandler(IBetRepository betRepository, ICompleteBetOutputPort outputPort) : IRequestHandler<CompleteBetRequest>
{
    private readonly IBetRepository betRepository = betRepository;
    private readonly ICompleteBetOutputPort outputPort = outputPort;

    public async Task Handle(CompleteBetRequest request, CancellationToken cancellationToken)
    {
        if (IsProofIsRequired(request))
        {
            outputPort.ProofIsMissing(request.BetId);
            return;
        }
        await betRepository.CompleteBetAsync(request.BetId, request.IsSuccess, request.Proof, cancellationToken);
        outputPort.Success(request.BetId, request.IsSuccess);
    }

    private static bool IsProofIsRequired(CompleteBetRequest request)
    {
        return request.IsSuccess && string.IsNullOrWhiteSpace(request.Proof);
    }
}
